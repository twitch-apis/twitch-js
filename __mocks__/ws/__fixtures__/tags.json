{
  "CLEARCHAT": [
    ":tmi.twitch.tv CLEARCHAT #dallas :ronni",
    "@ban-reason=Follow\\sthe\\srules :tmi.twitch.tv CLEARCHAT #dallas :ronni"
  ],
  "GLOBALUSERSTATE": "@color=#0D4200;display-name=dallas;emote-sets=0,33,50,237,793,2126,3517,4578,5569,9400,10337,12239;turbo=0;user-id=1337;user-type=admin :tmi.twitch.tv GLOBALUSERSTATE",
  "PRIVMSG": {
    "NON_BITS": "@badges=global_mod/1,turbo/1;color=#0D4200;display-name=dallas;emotes=25:0-4,12-16/1902:6-10;id=b34ccfc7-4977-403a-8a94-33c6bac34fb8;mod=0;room-id=1337;subscriber=0;tmi-sent-ts=1507246572675;turbo=1;user-id=1337;user-type=global_mod :ronni!ronni@ronni.tmi.twitch.tv PRIVMSG #dallas :Kappa Keepo Kappa",
    "BITS": "@badges=staff/1,bits/1000;bits=100;color=;display-name=dallas;emotes=;id=b34ccfc7-4977-403a-8a94-33c6bac34fb8;mod=0;room-id=1337;subscriber=0;tmi-sent-ts=1507246572675;turbo=1;user-id=1337;user-type=staff :ronni!ronni@ronni.tmi.twitch.tv PRIVMSG #dallas :cheer100"
  },
  "ROOMSTATE": {
    "JOIN": "@broadcaster-lang=en;r9k=0;slow=0;subs-only=0 :tmi.twitch.tv ROOMSTATE #dallas",
    "UPDATE_SLOW": "@slow=10 :tmi.twitch.tv ROOMSTATE #dallas"
  },
  "USERNOTICE": {
    "SUBSCRIPTION": "@badges=staff/1,broadcaster/1,turbo/1;color=#008000;display-name=ronni;emotes=;id=db25007f-7a18-43eb-9379-80131e44d633;login=ronni;mod=0;msg-id=sub;msg-param-months=1;msg-param-sub-plan=Prime;msg-param-sub-plan-name=Prime;room-id=1337;subscriber=1;system-msg=ronni\\sjust\\ssubscribed\\swith\\sTwitch\\sPrime!;tmi-sent-ts=1507246572675;turbo=1;user-id=1337;user-type=staff :tmi.twitch.tv USERNOTICE #dallas :Great stream -- keep it up!",
    "RESUBSCRIPTION": "@badges=staff/1,broadcaster/1,turbo/1;color=#008000;display-name=ronni;emotes=;id=db25007f-7a18-43eb-9379-80131e44d633;login=ronni;mod=0;msg-id=resub;msg-param-months=6;msg-param-sub-plan=Prime;msg-param-sub-plan-name=Prime;room-id=1337;subscriber=1;system-msg=ronni\\shas\\ssubscribed\\sfor\\s6\\smonths!;tmi-sent-ts=1507246572675;turbo=1;user-id=1337;user-type=staff :tmi.twitch.tv USERNOTICE #dallas :Great stream -- keep it up!",
    "SUBSCRIPTION_GIFT": "@badges=staff/1,premium/1;color=#0000FF;display-name=TWW2;emotes=;id=e9176cd8-5e22-4684-ad40-ce53c2561c5e;login=tww2;mod=0;msg-id=subgift;msg-param-months=1;msg-param-recipient-display-name=Mr_Woodchuck;msg-param-recipient-id=89614178;msg-param-recipient-name=mr_woodchuck;msg-param-sub-plan-name=House\\sof\\sNyoro~n;msg-param-sub-plan=1000;room-id=19571752;subscriber=0;system-msg=TWW2\\sgifted\\sa\\sTier\\s1\\ssub\\sto\\sMr_Woodchuck!;tmi-sent-ts=1521159445153;turbo=0;user-id=13405587;user-type=staff :tmi.twitch.tv USERNOTICE #forstycup",
    "RAID": "@badges=turbo/1;color=#9ACD32;display-name=TestChannel;emotes=;id=3d830f12-795c-447d-af3c-ea05e40fbddb;login=testchannel;mod=0;msg-id=raid;msg-param-displayName=TestChannel;msg-param-login=testchannel;msg-param-viewerCount=15;room-id=56379257;subscriber=0;system-msg=15\\sraiders\\sfrom\\sTestChannel\\shave\\sjoined\\n!;tmi-sent-ts=1507246572675;tmi-sent-ts=1507246572675;turbo=1;user-id=123456;user-type= :tmi.twitch.tv USERNOTICE #othertestchannel",
    "RITUAL": "@badges=;color=;display-name=SevenTest1;emotes=30259:0-6;id=37feed0f-b9c7-4c3a-b475-21c6c6d21c3d;login=seventest1;mod=0;msg-id=ritual;msg-param-ritual-name=new_chatter;room-id=6316121;subscriber=0;system-msg=Seventoes\\sis\\snew\\shere!;tmi-sent-ts=1508363903826;turbo=0;user-id=131260580;user-type= :tmi.twitch.tv USERNOTICE #seventoes :HeyGuys",
    "SUBMYSTERYGIFT": "@badges=subscriber/12,bits/5000,sub-gifter/1;color=;display-name=Nefarious_Ned;emotes=;id=761a4698-b867-4668-9658-c3844c7b8637;login=nefarious_ned;mod=0;msg-id=submysterygift;msg-param-mass-gift-count=5;msg-param-sender-count=37;msg-param-sub-plan=1000;room-id=73625408;subscriber=1;system-msg=Nefarious_Ned\\sis\\sgifting\\s5\\sTier\\s1\\sSubs\\sto\\sKate's\\scommunity!\\sThey've\\sgifted\\sa\\stotal\\sof\\s37\\sin\\sthe\\schannel!;tmi-sent-ts=1533875190783;turbo=0;user-id=109667930;user-type= :tmi.twitch.tv USERNOTICE #kate",
    "UNKNOWN": "@badges=;color=;display-name=SevenTest1;emotes=30259:0-6;id=37feed0f-b9c7-4c3a-b475-21c6c6d21c3d;login=seventest1;mod=0;msg-id=unknown;msg-param-unknown=value;room-id=6316121;subscriber=0;system-msg=Unknown\\ssystem\\smessage;tmi-sent-ts=1508363903826;turbo=0;user-id=131260580;user-type= :tmi.twitch.tv USERNOTICE #seventoes :Unknown USERNOTICE"
  },
  "USERSTATE": {
    "JOIN": "@color=#0D4200;display-name=ronni;emote-sets=0,33,50,237,793,2126,3517,4578,5569,9400,10337,12239;mod=1;subscriber=1;turbo=1;user-type=staff :tmi.twitch.tv USERSTATE #dallas"
  }
}