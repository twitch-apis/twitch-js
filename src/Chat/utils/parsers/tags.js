import { camelCase, gt, toLower, toUpper } from 'lodash'

import * as constants from '../../constants'
import * as types from './types'

const clearChat = tags => ({
  /**
   * CLEARCHAT tags
   * @typedef {Object} ClearChatTags
   * @property {string} [banReason]
   * @property {number} [banDuration]
   * @see https://dev.twitch.tv/docs/irc/tags#clearchat-twitch-tags
   */
  ...tags,
  banReason: types.generalString(tags.banReason),
  banDuration: types.generalNumber(tags.banDuration),
})

const globalUserState = tags => ({
  /**
   * GLOBALUSERSTATE tags
   * @typedef {Object} GlobalUserStateTags
   * @property {Array<string>} emoteSets
   * @property {string} userType
   * @property {string} username
   * @see https://dev.twitch.tv/docs/irc/tags#globaluserstate-twitch-tags
   */
  ...tags,
  emoteSets: types.emoteSets(tags.emoteSets),
  userType: types.userType(tags.userType),
  username: toLower(tags.displayName),
})

const privateMessage = (...args) =>
  userState(
    /** PRIVMSG tags
     * @typedef {UserStateTags} PrivateMessageTags
     * @see https://dev.twitch.tv/docs/irc/tags#privmsg-twitch-tags
     */
    ...args,
  )

const privateMessageCheerEvent = tags => {
  return gt(tags.bits, 0)
    ? { event: constants.EVENTS.CHEER, bits: parseInt(tags.bits, 10) }
    : {}
}

const roomState = roomStateTags =>
  /**
   * ROOMSTATE Tag
   * @typedef {Object} RoomStateTags
   * @property {string} [broadcasterLang]
   * @property {boolean} emoteOnly
   * @property {boolean|number} followersOnly
   * @property {boolean} r9k
   * @property {number} slow
   * @property {boolean} subsOnly
   * @see https://dev.twitch.tv/docs/irc/tags#roomstate-twitch-tags
   */
  Object.entries(roomStateTags).reduce((tags, [tag, value]) => {
    switch (tag) {
      case 'followersOnly':
        return { ...tags, [tag]: types.followersOnly(value) }
      // Strings
      case 'broadcasterLang':
        return { ...tags, [tag]: types.generalString(value) }
      // Numbers
      case 'slow':
        return { ...tags, [tag]: types.generalNumber(value) }
      // Booleans
      case 'emoteOnly':
      case 'r9k':
      case 'subsOnly':
        return { ...tags, [tag]: types.generalBoolean(value) }
      default:
        return { ...tags, [tag]: value }
    }
  }, {})

const userNotice = (...args) =>
  userState(
    /** USERNOTICE tags
     * @typedef {UserStateTags} UserNoticeTags
     * @see https://dev.twitch.tv/docs/irc/tags#usernotice-twitch-tags
     */
    ...args,
  )

const userNoticeMessageParameters = tags =>
  Object.entries(tags).reduce((parameters, [tag, value]) => {
    const [, param] = constants.MESSAGE_PARAMETER_PREFIX_RE.exec(tag) || []

    switch (param) {
      // Numbers.
      case 'Months':
      case 'MassGiftCount':
      case 'PromoGiftTotal':
      case 'SenderCount':
      case 'ViewerCount':
        return { ...parameters, [camelCase(param)]: types.generalNumber(value) }
      // Not a msgParam.
      case undefined:
        return parameters
      // Strings
      default:
        return { ...parameters, [camelCase(param)]: types.generalString(value) }
    }
  }, {})

const userNoticeEvent = tags => {
  switch (tags.msgId) {
    case constants.USER_NOTICE_MESSAGE_IDS.ANON_GIFT_PAID_UPGRADE:
      return constants.EVENTS.ANON_GIFT_PAID_UPGRADE
    case constants.USER_NOTICE_MESSAGE_IDS.GIFT_PAID_UPGRADE:
      return constants.EVENTS.GIFT_PAID_UPGRADE
    case constants.USER_NOTICE_MESSAGE_IDS.RESUBSCRIPTION:
      return constants.EVENTS.RESUBSCRIPTION
    case constants.USER_NOTICE_MESSAGE_IDS.RAID:
      return constants.EVENTS.RAID
    case constants.USER_NOTICE_MESSAGE_IDS.RITUAL:
      return constants.EVENTS.RITUAL
    case constants.USER_NOTICE_MESSAGE_IDS.SUBSCRIPTION:
      return constants.EVENTS.SUBSCRIPTION
    case constants.USER_NOTICE_MESSAGE_IDS.SUBSCRIPTION_GIFT:
      return constants.EVENTS.SUBSCRIPTION_GIFT
    case constants.USER_NOTICE_MESSAGE_IDS.SUBSCRIPTION_GIFT_COMMUNITY:
      return constants.EVENTS.SUBSCRIPTION_GIFT_COMMUNITY
    default:
      return toUpper(tags.msgId)
  }
}

const userState = tags => ({
  /**
   * USERSTATE tags
   * @typedef {Object} UserStateTags
   * @property {BadgesTag} badges
   * @property {Array<EmoteTag>} emotes
   * @property {Array<string>} emoteSets
   * @property {number} [bits]
   * @see https://dev.twitch.tv/docs/irc/tags#userstate-twitch-tags
   */
  ...tags,
  badges: types.badges(tags.badges),
  bits: types.generalNumber(tags.bits),
  emotes: types.emotes(tags.emotes),
  emoteSets: types.emoteSets(tags.emoteSets),
  userType: types.userType(tags.userType),
})

export {
  clearChat,
  globalUserState,
  privateMessage,
  privateMessageCheerEvent,
  roomState,
  userNotice,
  userNoticeMessageParameters,
  userNoticeEvent,
  userState,
}
