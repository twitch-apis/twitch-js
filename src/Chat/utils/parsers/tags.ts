import { camelCase, gt, toLower, toUpper } from 'lodash'

import * as constants from '../../constants'
import * as types from './types'

/**
 * CLEARCHAT tags
 * @typedef {Object} ClearChatTags
 * @property {string} [banReason]
 * @property {number} [banDuration]
 * @see https://dev.twitch.tv/docs/irc/tags#clearchat-twitch-tags
 */
const clearChat = tags => ({
  ...tags,
  banReason: types.generalString(tags.banReason),
  banDuration: types.generalNumber(tags.banDuration),
})

/**
 * GLOBALUSERSTATE tags
 * @typedef {Object} GlobalUserStateTags
 * @property {Array<string>} emoteSets
 * @property {string} userType
 * @property {string} username
 * @see https://dev.twitch.tv/docs/irc/tags#globaluserstate-twitch-tags
 */
const globalUserState = tags => ({
  ...tags,
  emoteSets: types.emoteSets(tags.emoteSets),
  userType: types.userType(tags.userType),
  username: toLower(tags.displayName),
})

/** PRIVMSG tags
 * @typedef {UserStateTags} PrivateMessageTags
 * @see https://dev.twitch.tv/docs/irc/tags#privmsg-twitch-tags
 */
const privateMessage = (...args) => userState(...args)

const privateMessageCheerEvent = tags => {
  return gt(tags.bits, 0)
    ? { event: constants.EVENTS.CHEER, bits: parseInt(tags.bits, 10) }
    : {}
}

/**
 * ROOMSTATE Tag
 * @typedef {Object} RoomStateTags
 * @property {string} [broadcasterLang]
 * @property {booelan} emoteOnly
 * @property {boolean|number} followersOnly
 * @property {boolean} r9k
 * @property {number} slow
 * @property {boolean} subsOnly
 * @see https://dev.twitch.tv/docs/irc/tags#roomstate-twitch-tags
 */
const roomState = roomStateTags =>
  Object.entries(roomStateTags).reduce((tags, [tag, value]) => {
    switch (tag) {
      case 'followersOnly':
        return { ...tags, [tag]: types.followersOnly(value) }
      // Strings
      case 'broadcasterLang':
        return { ...tags, [tag]: types.generalString(value) }
      // Numbers
      case 'slow':
        return { ...tags, [tag]: types.generalNumber(value) }
      // Booleans
      case 'emoteOnly':
      case 'r9k':
      case 'subsOnly':
        return { ...tags, [tag]: types.generalBoolean(value) }
      default:
        return { ...tags, [tag]: value }
    }
  }, {})

/** USERNOTICE tags
 * @typedef {UserStateTags} UserNoticeTags
 * @see https://dev.twitch.tv/docs/irc/tags#usernotice-twitch-tags
 */
const userNotice = (...args) => userState(...args)

const userNoticeMessageParameters = tags =>
  Object.entries(tags).reduce((parameters, [tag, value]) => {
    const [, param] = constants.MESSAGE_PARAMETER_PREFIX_RE.exec(tag) || []

    switch (param) {
      // Numbers.
      case 'Months':
      case 'MassGiftCount':
      case 'PromoGiftTotal':
      case 'SenderCount':
      case 'ViewerCount':
        return { ...parameters, [camelCase(param)]: types.generalNumber(value) }
      // Not a msgParam.
      case undefined:
        return parameters
      // Strings
      default:
        return { ...parameters, [camelCase(param)]: types.generalString(value) }
    }
  }, {})

const userNoticeEvent = tags => {
  switch (tags.msgId) {
    case constants.USER_NOTICE_MESSAGE_IDS.ANON_GIFT_PAID_UPGRADE:
      return constants.EVENTS.ANON_GIFT_PAID_UPGRADE
    case constants.USER_NOTICE_MESSAGE_IDS.GIFT_PAID_UPGRADE:
      return constants.EVENTS.GIFT_PAID_UPGRADE
    case constants.USER_NOTICE_MESSAGE_IDS.RESUBSCRIPTION:
      return constants.EVENTS.RESUBSCRIPTION
    case constants.USER_NOTICE_MESSAGE_IDS.RAID:
      return constants.EVENTS.RAID
    case constants.USER_NOTICE_MESSAGE_IDS.RITUAL:
      return constants.EVENTS.RITUAL
    case constants.USER_NOTICE_MESSAGE_IDS.SUBSCRIPTION:
      return constants.EVENTS.SUBSCRIPTION
    case constants.USER_NOTICE_MESSAGE_IDS.SUBSCRIPTION_GIFT:
      return constants.EVENTS.SUBSCRIPTION_GIFT
    case constants.USER_NOTICE_MESSAGE_IDS.SUBSCRIPTION_GIFT_COMMUNITY:
      return constants.EVENTS.SUBSCRIPTION_GIFT_COMMUNITY
    default:
      return toUpper(tags.msgId)
  }
}

/**
 * USERSTATE tags
 * @typedef {Object} UserStateTags
 * @property {BadgesTag} badges
 * @property {Array<EmoteTag>} emotes
 * @property {Array<string>} emoteSets
 * @property {number} [bits]
 * @see https://dev.twitch.tv/docs/irc/tags#userstate-twitch-tags
 */
const userState = tags => ({
  ...tags,
  badges: types.badges(tags.badges),
  bits: types.generalNumber(tags.bits),
  emotes: types.emotes(tags.emotes),
  emoteSets: types.emoteSets(tags.emoteSets),
  userType: types.userType(tags.userType),
})

export {
  clearChat,
  globalUserState,
  privateMessage,
  privateMessageCheerEvent,
  roomState,
  userNotice,
  userNoticeMessageParameters,
  userNoticeEvent,
  userState,
}
