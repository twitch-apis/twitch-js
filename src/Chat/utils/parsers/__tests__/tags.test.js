import * as tags from '../tags'

describe('Chat/utils/parsers/tags', () => {
  describe('privateMessageCheerEvent', () => {
    test('should return empty object for non-bits message', () => {
      const actual = tags.privateMessageCheerEvent({ bits: 0 })
      const expected = {}

      expect(actual).toEqual(expected)
    })
    test('should return cheer event object', () => {
      const actual = tags.privateMessageCheerEvent({ bits: 100 })

      expect(actual).toMatchSnapshot()
    })
  })
})
