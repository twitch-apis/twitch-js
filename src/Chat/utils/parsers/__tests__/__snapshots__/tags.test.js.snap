// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat/utils/parsers/tags privateMessageCheerEvent should return cheer event object 1`] = `
Object {
  "bits": 100,
  "event": "CHEER",
}
`;
